feat: Show completed recipe view

ffaass opened the pull request:
Resolves: #8

ffaass committed:
feat: Add recipe view template

ffaass committed:
refactor: Add recipe model key to Template constant class

ffaass committed:
feat: Divide recipe template by complete state

ffaass committed:
refactor: Optimize import

ffaass committed:
feat: Change step rendering policy

- Exclude closed step when render completed recipe

ffaass committed:
Merge branch 'develop' into feature/show-complete-recipe

# Conflicts:
#	src/main/java/com/woowahan/techcamp/recipehub/recipe/service/RecipeService.java
#	src/main/java/com/woowahan/techcamp/recipehub/recipestep/domain/AbstractRecipeStep.java
#	src/main/java/com/woowahan/techcamp/recipehub/recipestep/domain/RecipeStep.java
#	src/main/resources/templates/base.hbs
#	src/test/java/com/woowahan/techcamp/recipehub/recipe/controller/RecipeAcceptanceTest.java
#	src/test/java/com/woowahan/techcamp/recipehub/recipe/service/RecipeServiceTest.java

ffaass committed:
feat: Add recipe detail view style

- Use bulma framework
- Move head tag to outside of {{#partial "head"}} in base.hbs
- Add incomplete recipe detail view template
  - 현재는 내용이 크게 다르지 않으나 미완성 레시피 뷰 작업할 때 많이 달라질 것
- Modity & Add local dummy data set

CHANMYUNG commented on src/main/java/com/woowahan/techcamp/recipehub/RecipehubApplication.java:
```diff
@@ -6,7 +6,7 @@
-    public static void main(String[] args) {
-        SpringApplication.run(RecipehubApplication.class, args);
-    }
+	public static void main(String[] args) {
```
6 spaces indentation이네요?

ryuneeee replied:
이거 아마 Tab이 Spacing으로 안되는거 같아요. \t 문자로 보이네요

CHANMYUNG commented on src/main/java/com/woowahan/techcamp/recipehub/recipestep/domain/AbstractRecipeStep.java:
```diff
@@ -3,14 +3,18 @@
 @Table(name = "RecipeStep")
 @Inheritance(strategy = InheritanceType.SINGLE_TABLE)
-@DiscriminatorColumn(name = "type")
+@DiscriminatorColumn(name="type")
```
여기 왜? 혹시 reformat code하고 optimize import 안하셨나용??

CHANMYUNG reviewed (changes requested):
구현 잘 했네요. 리뷰했어요.

Lahavas commented on src/main/java/com/woowahan/techcamp/recipehub/common/Template.java:
```diff
@@ -0,0 +1,10 @@
+package com.woowahan.techcamp.recipehub.common;
+
+public class Template {
+    public static final String RECIPE_COMPLETED = "/recipe-completed";
```
여기 클래스명이 살짝 모호하다고 생각됩니다. 추후에 팀원들과 함께 논의해볼 필요가 있을거같습니다~

Lahavas commented on src/main/resources/data-local.sql:
```diff
@@ -2,6 +2,12 @@ INSERT INTO user (email, password, name) VALUES ('test@er.com', '$2a$10$6dbJkT99
+INSERT INTO recipe (owner_id, name, completed, img_url, created_at, updated_at) VALUES (1, '로즈마리 후추 버거', false, 'https://cdn.bmf.kr/_data/product/I21A3/a5ae10184ec276667e0a35e6f3012f20.jpg', now(), now());
+INSERT INTO recipe_step (type, content, img_url, name, closed, sequence, rejected, recipe_id, writer_id) VALUES ('Step', '[''빵을 얹는다'', ''고기를 올린다'']', 'https://cdn.bmf.kr/_data/product/IF088/2dee3415155804a5e9dc7c24fd21d96c.jpg', '버거 모양 만들기', false, 1, false, 2, 1);
+INSERT INTO recipe_step (type, content, img_url, name, closed, sequence, rejected, recipe_id, writer_id) VALUES ('Step', '[''빵을 얹는다'']', 'https://cdn.bmf.kr/_data/product/IF088/2dee3415155804a5e9dc7c24fd21d96c.jpg', '버거 모양 만들기2', false, 2, false, 2, 2);
+INSERT INTO recipe_step (type, content, img_url, name, closed, sequence, rejected, recipe_id, writer_id) VALUES ('Step', '[''빵을 얹는다'']', 'https://cdn.bmf.kr/_data/product/IF088/2dee3415155804a5e9dc7c24fd21d96c.jpg', '버거 모양 만들기3', false, 3, false, 2, 3);
```
위의 recipe & recipe_step 더미데이터랑 잘 구분이 안되는거같습니다.

ryuneeee replied:
Paragraph divide용 SQL 주석 달아주시면 감사!

Lahavas commented on src/main/resources/templates/index.hbs:
```diff
@@ -1,44 +1,44 @@
-                            <p class="subtitle is-6">{{owner.name}}</p>
-                        </div>
+<section class="section">
+    <div class="container">
```
여기 reformat code 잘못 먹힌거같네요 Indent가 잘못 들어갔습니다~

Lahavas commented on src/main/resources/templates/recipe-completed.hbs:
```diff
@@ -0,0 +1,33 @@
+{{#partial "contents"}}
+{{#recipe}}
+<div class="section">
+    <div class="container">
```
base.hbs에 contents 바깥 부분에 body Tag가 정의되어있지 않아서 이 부분은 body Tag로 감싸줘야할거같네요~

Lahavas commented on src/main/resources/templates/recipe-incompleted.hbs:
```diff
@@ -0,0 +1,34 @@
+{{#partial "contents"}}
+{{#recipe}}
+<div class="section">
+    <div class="container">
```
base.hbs에 contents 바깥 부분에 body Tag가 정의되어있지 않아서 이 부분은 body Tag로 감싸줘야할거같네요~

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/recipe/domain/Recipe.java:
```diff
@@ -55,4 +57,10 @@ public Recipe(Category category, User owner, String name, boolean completed, Dat
     }
+
+    public void removeClosedSteps() {
+        this.recipeSteps = this.recipeSteps.stream()
```
기존 Variable에 덮어쓰는건 좋지 않아요~

ryuneeee replied:
아, 구조상 그런거구나. 이것도 논의를 하는 방향으로 가볼까요

Lahavas replied:
저도 이부분 고민되더라구요 논의해보져~

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/recipe/service/RecipeService.java:
```diff
@@ -7,19 +7,27 @@
-
     @Autowired
-    RecipeRepository recipeRepository;
+    private RecipeRepository recipeRepository;
```
제가 놓친거 잘 챙겨주셨네 ㅋㅋㅋㅋ🤩

ryuneeee commented on src/main/java/com/woowahan/techcamp/recipehub/recipe/service/RecipeService.java:
```diff
@@ -7,19 +7,27 @@
+    public Recipe findById(Long recipeId) {
+        Recipe recipe = recipeRepository.findById(recipeId).orElseThrow(EntityNotFoundException::new);
+        if (recipe.isCompleted()) {
+            recipe.removeClosedSteps();
```
Entity를 find를 하는 과정에서, Entity를 수정하는건 좀 이상한거 같아요🧐

ryuneeee reviewed (changes requested):
findById만 수정 해 주시면 될듯😇

Lahavas commented on src/main/java/com/woowahan/techcamp/recipehub/recipe/controller/RecipeController.java:
```diff
@@ -0,0 +1,29 @@
+
+@Controller
+@RequestMapping("/recipe")
+public class RecipeController {
```
여기 개인적인 의견으로는.... recipe보단 recipes로 Routing을 지정하는게 더 좋을거같네요

crongro commented:
오 엄격한 리뷰!

ffaass committed:
refactor: Move removing closed steps logic in domain object to new DTO

- Create RecipeDetailDTO
- Rename RecipeDTO to RecipeCreationDTO
- Move removeClosedSteps method to DTO creation logic with test code
- Remove Template class
  - Because of useless and conflict issue

ffaass committed:
refactor: Change recipe controllers' path mapping

- Change /recipe to /recipes
- Add link of detail page to recipe content card in index.hbs

ffaass committed:
chore: Rearrange local dummy data

- Add comment
- Change field order

ffaass committed:
Merge branch 'develop' into feature/show-complete-recipe

minzzzy commented on src/main/java/com/woowahan/techcamp/recipehub/recipe/dto/RecipeCreationDTO.java:
```diff
@@ -24,7 +24,7 @@
 
     @Builder
-    public RecipeDTO(@NotNull Category category, @NotEmpty @NotNull String name, String imgUrl) {
+    public RecipeCreationDTO(@NotNull Category category, @NotEmpty @NotNull String name, String imgUrl) {
```
생성자를 만들어 줄 때도 NotNull과 같은 annotation을 붙여줘야 하나요? 불필요하면 빼줘도 좋을 것 같습니다.

minzzzy reviewed (approved):
저는 승인합니다~

CHANMYUNG reviewed (approved):
제가 이슈 쳐내기 바빠서요 ㅎㅎ;; 
양심선언합니다.
"저 바뀐 코드 리뷰 안했어요"

CHANMYUNG commented:
컨플릭트 뭔가요!!

ffaass committed:
Merge branch 'develop' into feature/show-complete-recipe

Lahavas reviewed (approved):
승인할게요
